### ISSUE TITLE
`after_commit` callback?

### ISSUE THREAD

**Issue by jagthedrummer** [JSONAPI-Resources/jsonapi-resources#503]
I'm running into an issue where I'm sending a job to RabbitMQ in an `after_create` callback on a resource, and sometimes the job gets picked up for work before the transaction to create the record has committed. Is there anything analogous to an `after_commit` callback?

**Comment by lgebhardt** [comment]
I think `after_operations` would be what you want. The `operations` callbacks wrap the transaction.

**Comment by jagthedrummer** [comment]
Thanks for the pointer. If I'm reading the docs right it looks like `after_operations` is only available on an `OperationsProcessor` and not on a `Resource`. Is that right? If so, is there a way to specify an `OperationsProcessor` to use for just one particular `Resource`? (I don't need to send a message for every operation, only for `create` operations on a `Profile` resource.)

**Comment by lgebhardt** [comment]
You are correct that the `after_operations` callback is on the OperationProcessor. The issue is that eventually (once the bulk spec is completed) we will support multiple operations which will all be committed at once. So having the callback on the individual resources won't be easy (though I'm assuming we could probably make it work).

This should get you close, but the code hasn't been tested:

``` ruby
class MyOperationsProcessor < ActiveRecordOperationsProcessor
  after_operations :send_to_rabbitmq

  def send_to_rabbitmq
    if @operation.resource_klass.is_a?(FooResource) && @operation.is_a?(CreateResourceOperation)
      #do the rabbit stuff here
    end
  end
end
```

**Comment by jagthedrummer** [comment]
Awesome! Thanks for pointing me in the right direction.

**Comment by xtian** [comment]
Just wanted to chime in and say that we're having this problem with Sidekiq as well. It would be great to have this callback available on the resource itself since we have disparate resource-specific logic that I don't think makes a lot of sense to put into a Processor.

**Comment by lgebhardt** [comment]
@xtian I understand. I'll give some thought to possible solutions.

**Comment by bartoszkopinski** [comment]
+1 to after_commit & after_rollback callbacks

**Comment by mctaylorpants** [comment]
+1 to `after_commit` as well, we had a similar issue where we wanted to run a service after the transaction had closed. We used `@operation.is_a?(CreateResourceOperation)` and it worked for now, but it would be great to have a more robust way of doing this 😄

**Comment by hidde-jan** [comment]
Any progress on this? How do you schedule jobs in your JR projects, @lgebhardt?

Looking at https://github.com/cerebris/jsonapi-resources/blob/e53a32daa4bde4c2e2120d1abf55644a9a060f23/lib/jsonapi/operation_dispatcher.rb#L13, it seems that there is no longer a way to do something after the transaction is committed. Is that right?

**Comment by daniely** [comment]
Should this issue be closed due to the way [operational processors](http://jsonapi-resources.com/v0.8/guide/operation_processors.html) work in v0.8?

From the guide:

> The processor is created based on the resource name, including the namespace. If a processor does not exist for a resource (namespace matters) the default operation processor is used instead.

**Comment by hidde-jan** [comment]
I don't think so. The processor still runs within a single transaction. The after commit callback would be for when you are 100% sure that the changes were committed to the database.

**Comment by ghost** [comment]
The `after_operations` callback is gone from processors, with no callbacks being run outside of the transaction (in 0.8.1). Is there a way to do this now?

**Comment by ghost** [comment]
Ah wait the `after_operations` callback is on the controller now.

**Comment by ghost** [comment]
For those who are finding this via Google or searching for `after_commit` here is a work around:

```ruby
class SomeController < ApplicationController
  include JSONAPI::ActsAsResourceController

  after_process_operations :do_something

  def do_something
    if @request.operations.map(&:operation_type).include?(:create_resource)
      unless @operation_results.has_errors?
        resource = @operation_results.results.first.resource

        # do more stuff with resource
      end
    end
  end
end
```

**Comment by bartoszkopinski** [comment]
"Operations" callback looks more like a hack than the proper way to do it. Can't believe it's the recommended solution.

This is what I ended up doing (feels way nicer and allows you to schedule background jobs anywhere within transactions - not only for jsonapi-resources):
```ruby
class AsyncRecord
  def initialize klass, *args
    @klass = klass
    @args = args
  end

  def has_transactional_callbacks?
    true
  end

  def committed!(*)
    enqueue
  rescue => ex
    Rails.logger.warn("Transaction commited - async scheduling failed")
  end

  def rolledback!(*_, **__)
    Rails.logger.warn("Transaction rolledback! - async scheduling skipped")
  end

  def enqueue
    if klass < Sidekiq::Worker
      klass.perform_async(*args)
    elsif klass < ActiveJob::Base
      klass.perform_later(*args)
    else
      klass.delay.send(:perform, *args)
    end
  end

  private
  attr_reader :args, :klass
end
```

```ruby
def safe_delay worker_klass, *args
  async_record = AsyncRecord.new(worker_klass, *args)
  if ActiveRecord::Base.connection.transaction_open?
    ActiveRecord::Base.connection.current_transaction.add_record(async_record)
  else
    async_record.enqueue
  end
end
```
Usage:
```ruby
safe_delay(BackgroundWorker, @model.id)
```

Based on:
http://blog.arkency.com/2015/10/run-it-in-background-job-after-commit/

**Comment by ramontayag** [comment]
https://github.com/grosser/ar_after_transaction might be useful. You should be able to use this in an after create callback

**Comment by hidde-jan** [comment]
I again ran into a race condition with our worker backend due to this missing feature. We shouldn't have to use some workaround. This should be easy to implement directly on a resource, not on an operation or controller or processor (ok maybe there).

**Comment by krainboltgreene** [comment]
Just an FYI, raising an exception in an `after_process_operations` doesn't actually raise an error in as a response.
